qiskit_experiments/analysis/curve_fitting.py

```diff
@@ -0,0 +1,220 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Curve fitting functions for experiment analysis
+"""
+# pylint: disable = invalid-name
+
+from typing import List, Dict, Tuple, Callable, Optional
+
+import numpy as np
+import scipy.optimize as opt
+
+from qiskit.exceptions import QiskitError
+from qiskit_experiments.base_analysis import AnalysisResult
+from qiskit_experiments.analysis.data_processing import filter_data
+
+
+def curve_fit(
+    func: Callable,
+    xdata: np.ndarray,
+    ydata: np.ndarray,
+    p0: np.ndarray,
+    sigma: Optional[np.ndarray] = None,
+    **kwargs,
+) -> AnalysisResult:
+    """Use non-linear least squares to fit a function, f, to data
+
+    Wraps scipy.optimize.curve_fit
+
+    Args:
+        func: a fit function `f(x *params)`.
+        xdata: a 1D array of x-data
+        ydata: a 1D array of y-data
+        p0: initial guess for optimization parameters.
+        sigma: Optional, a 1D array of standard deviations in ydata.
+        kwargs: additional kwargs for scipy.optimize.curve_fit.
+
+    Returns:
+        AnalysisResult: result containing `popt` the optimal fit parameters,
+                        `popt_err` the standard error estimates popt,
+                        `pcov` the covariance matrix for the fit,
+                        `chisq` the chi-squared parameter of fit,
+                        `xrange` the range of xdata values used for fit.
+    """
+
+    # Run curve fit
+    # pylint: disable = unbalanced-tuple-unpacking
+    popt, pcov = opt.curve_fit(func, xdata, ydata, sigma=sigma, p0=p0, **kwargs)
+    popt_err = np.sqrt(np.diag(pcov))
+
+    # Compute chi-squared for fit
+    yfits = func(xdata, *popt)
+    chisq = np.mean(((yfits - ydata) / sigma) ** 2)
+
+    # Compute xdata range for fit
+    xdata_range = [min(xdata), max(xdata)]
```

Comment:
Can we optionally truncate the fit range? i.e. adding range truncation function in the helper library. This is useful for fitting power dependency (sometime we see non-linearly with large xvalue, and want to exclude this region)

Reply:
I could add something like a `xbounds=[xmin, xmax]` kwarg to the curve_fit_data function.

Or maybe a more general solution would be to beef up the filters so they aren't just value checks `meta_val == filter_val`, but instead callables `filter_data(data, meta_key=fn)` where  the fn is callable that evaluates to a bool `fn(meta_val) -> bool`.

Im not sure how you could allow both callable and fixed value filters though. They might just have to be all callable if we do that (like a fixed value one would need to be done something like `filter_data(data, meta_key=(lambda x: x == target))`

Reply:
I think **filter is used to filter out circuit based on metadata so current implementation looks good. `xbounds` kwarg looks good to me. I also noticed the function lacks `bounds` for parameter (if this is scipy wrapper we should have this argument).

Reply:
Any of scipy `curve_fits` other kwargs such as `bounds` are passed through via the **kwargs of `curve_fit` and `multi_curve_fit`

Reply:
yeah I just meant consistency of function signature, or docstring. current interface is ok in terms of functionality.